engine/table/src/main/java/io/deephaven/engine/util/BigDecimalUtils.java

```diff
@@ -44,67 +52,148 @@ public PrecisionAndScale(final int precision, final int scale) {
                 }
+            } else {
+                throw new IllegalArgumentException("Column source is not of type BigDecimal or an array/vector of " +
+                        "BigDecimal, but of type " + columnType + " and component type " +
+                        columnSource.getComponentType());
             }
         }
 
-        // If these are < 0, then every value we visited was null
-        if (maxPrecisionMinusScale < 0 && maxScale < 0) {
+        // If these are same as initial values, then every value we visited was null
+        if (result.maxPrecisionMinusScale == INIT_MAX_PRECISION_MINUS_SCALE && result.maxScale == INIT_MAX_SCALE) {
             return EMPTY_TABLE_PRECISION_AND_SCALE;
         }
 
-        return new PrecisionAndScale(maxPrecisionMinusScale + maxScale, maxScale);
+        return new PrecisionAndScale(result.maxPrecisionMinusScale + result.maxScale, result.maxScale);
+    }
+
+    private static class ComputationResult {
+        private int maxPrecisionMinusScale;
+        private int maxScale;
+
+        private ComputationResult(final int maxPrecisionMinusScale, final int maxScale) {
+            this.maxPrecisionMinusScale = maxPrecisionMinusScale;
+            this.maxScale = maxScale;
+        }
+    }
+
+    private static void processFlatColumn(
+            @NotNull final ColumnSource<?> columnSource,
+            @NotNull final RowSequence.Iterator it,
+            @NotNull final ChunkSource.GetContext context,
+            @NotNull final ComputationResult result) {
+        while (it.hasMore()) {
+            final RowSequence rowSeq = it.getNextRowSequenceWithLength(TARGET_CHUNK_SIZE);
+            final ObjectChunk<BigDecimal, ? extends Values> chunk =
+                    columnSource.getChunk(context, rowSeq).asObjectChunk();
+            final int numValues = chunk.size();
+            for (int i = 0; i < numValues; ++i) {
+                final BigDecimal value = chunk.get(i);
+                updateMaximum(result, value);
+            }
+        }
+    }
+
+    private static void processVectorColumn(
+            @NotNull final ColumnSource<?> columnSource,
+            @NotNull final RowSequence.Iterator it,
+            @NotNull final ChunkSource.GetContext context,
+            @NotNull final ComputationResult result) {
+        while (it.hasMore()) {
+            final RowSequence rowSeq = it.getNextRowSequenceWithLength(TARGET_CHUNK_SIZE);
+            final ObjectChunk<ObjectVector<BigDecimal>, ? extends Values> chunk =
+                    columnSource.getChunk(context, rowSeq).asObjectChunk();
+            final int numValues = chunk.size();
+            for (int i = 0; i < numValues; ++i) {
+                final ObjectVector<BigDecimal> values = chunk.get(i);
+                if (values == null) {
+                    continue;
+                }
+                for (final BigDecimal value : values) {
+                    updateMaximum(result, value);
+                }
```

Comment:
I prefer explicitly ensuring we close the iterator. The previous pattern relied on our "close upon exhaustion" behavior.
```suggestion
                try (final CloseableIterator<BigDecimal> valuesIterator = values.iterator()) {
                    valuesIterator.forEachRemaining(v -> updateMaximum(result, v));
                }
```